src/MessagingClient.ts

```diff
@@ -402,7 +402,7 @@ export class MessagingClient implements MessagingAPI {
                 })
                 publicRooms.push(...res.chunk)
                 nextBatch = res.next_batch
-            } while (publicRooms.length < limit || !res.next_batch)
+            } while (publicRooms.length < limit && res.next_batch)
```

Comment:
Looks good. Let's add an integration test to reproduce this issue and check the solution.